setup.py

```diff
@@ -4,6 +4,39 @@
 
 from setuptools import find_packages, setup
 
+REQUIRED_PYTHON = (3, 5)
+
+
+# Fail with explicit error if trying to install Django on an incompatible
+# (too old) version of Python, and advise on how to avoid this.
+HAS_PYTHON = sys.version_info[:2]
+if HAS_PYTHON < REQUIRED_PYTHON:
+    sys.stderr.write("""
+
+===========
+FATAL ERROR
+===========
+
+This version of Django requires Python {}, but you are trying to
+install it on Python {}.
+
+This is probably because you are using an old version of pip, which
+does not understand the python_requires classifier. Make sure you
+have pip >= 9.0 and setuptools >= 24.2, then try again:
```

Comment:
Any response to this comment about not needing setuptools here?

Reply:
The setuptools version is (in practice) not an issue:

In the rare case that someone force a source-install (by using the `--no-binary`) flag, then pip will get the sdist, try to build it, which will fail on too-old setuptools. 

This is/was more of an issue for project like numpy, that do not have universal wheels, and thus for most linux users where pip would get the sdist (by default of not finding a compatible wheel). Now that there is the manylinux1, it not really an issue. And anyway django have an universal wheel.

Python version check also happen before setuptools version matters, so users that have old setuptools may not even see that. 

So IMHO the setuptools is unnecessary.

The only potential help is the correlation between setuptools version and python version, so Python 2 users are more likely to have an old setuptools. I can't judge how frequent this may be in the Django community.

Reply:
Ah, sorry Tim - I'd taken it out, then read the notes on sdist installs [here](http://www.python3statement.org/practicalities/) and put it back in.

I've left it in because it may catch problems with eg `pip install -e .` from a git clone, and because once a user is updating pip it's no extra trouble to get setuptools too.